tests/integration_tests/core/test_lock_store.py

```diff
@@ -0,0 +1,194 @@
+import asyncio
+import time
+from pathlib import Path
+
+import numpy as np
+import pytest
+from _pytest.monkeypatch import MonkeyPatch
+
+from rasa.core.agent import Agent
+from rasa.core.channels import UserMessage
+from rasa.shared.constants import INTENT_MESSAGE_PREFIX
+from rasa.core.lock_store import LockError, RedisLockStore
+
+
+def test_create_lock_store(redis_lock_store: RedisLockStore):
+    conversation_id = "my id 0"
+
+    # create and lock
+    lock = redis_lock_store.create_lock(conversation_id)
+    redis_lock_store.save_lock(lock)
+    lock = redis_lock_store.get_lock(conversation_id)
+    assert lock
+    assert lock.conversation_id == conversation_id
+
+
+def test_serve_ticket(redis_lock_store: RedisLockStore):
+    conversation_id = "my id 1"
+
+    lock = redis_lock_store.create_lock(conversation_id)
+    redis_lock_store.save_lock(lock)
+
+    # issue ticket with long lifetime
+    ticket_0 = redis_lock_store.issue_ticket(conversation_id, 10)
+    assert ticket_0 == 0
+
+    lock = redis_lock_store.get_lock(conversation_id)
+    assert lock.last_issued == ticket_0
+    assert lock.now_serving == ticket_0
+    assert lock.is_someone_waiting()
+
+    # issue another ticket
+    ticket_1 = redis_lock_store.issue_ticket(conversation_id, 10)
+
+    # finish serving ticket_0
+    redis_lock_store.finish_serving(conversation_id, ticket_0)
+
+    lock = redis_lock_store.get_lock(conversation_id)
+
+    assert lock.last_issued == ticket_1
+    assert lock.now_serving == ticket_1
+    assert lock.is_someone_waiting()
+
+    # serve second ticket and no one should be waiting
+    redis_lock_store.finish_serving(conversation_id, ticket_1)
+
+    lock = redis_lock_store.get_lock(conversation_id)
+    assert not lock.is_someone_waiting()
+
+
+# noinspection PyProtectedMember
+def test_lock_expiration(redis_lock_store: RedisLockStore):
+    conversation_id = "my id 2"
+    lock = redis_lock_store.create_lock(conversation_id)
+    redis_lock_store.save_lock(lock)
+
+    # issue ticket with long lifetime
+    ticket = lock.issue_ticket(10)
+    assert ticket == 0
+    assert not lock._ticket_for_ticket_number(ticket).has_expired()
+
+    # issue ticket with short lifetime
+    ticket = lock.issue_ticket(0.00001)
+    time.sleep(0.00002)
+    assert ticket == 1
+    assert lock._ticket_for_ticket_number(ticket) is None
+
+    # newly assigned ticket should get number 1 again
+    assert lock.issue_ticket(10) == 1
+
+
+async def test_multiple_conversation_ids(default_agent: Agent):
```

Comment:
I don't understand this test. Seems to me it tests if `async.gather` works as expected?

Reply:
I think that the agent's `handle_message` leverages the lock store, so this test ensure it works correctly. I copy pasted this from the unit tests, guessing it could be useful to have an integration test for it. WDYT?

Reply:
I wonder if there is a way to have these tests in a test class and then just run them in different contexts e.g. inmemory for the unit tests, but with redis here?
Maybe a future improvement...

Reply:
🤔 yes could be a good idea at some point, seems a bit premature to me though.